crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -477,8 +477,9 @@ impl<'db> TypeInferenceBuilder<'db> {
         for class in class_definitions {
             // (1) Check that the class does not have a cyclic definition
             if class.is_cyclically_defined(self.db) {
-                self.diagnostics.add(
-                    class.node(self.db).into(),
+                self.diagnostics.add_if_definition_in_same_file(
+                    self.index,
+                    class.node(self.db),
```

Comment:
Other diagnostics might need the same "treatment". Again, I first wanted to see if this goes in the right direction. Maybe we want to store a reference to `SemanticIndexBuilder` in the `DiagnosticsBuilder` and check this for every diagnostic that we add?